### PR TITLE
Revamped the main configuration file

### PR DESCRIPTION
**Pull request by javiereguiluz**
The main configuration file is a mess and looks outdated. This PR revamps everything and merges the main 4 config files into a single reference file. This unique file explains 80% of what you need to know about config:

* The config/ structure
* The config/packages/ files
* Config formats
* Config parameters
* Config Environments
* Env Vars
* The multiple .env files

### CODE REVIEW

File: configuration.rst [hunk 1]
```diff
@@ -1,136 +1,158 @@
...
-The top-level key (here ``framework``) references configuration for a specific
-bundle (:doc:`FrameworkBundle </reference/configuration/framework>` in this case).
+Unlike other frameworks, Symfony doesn't impose you a specific format to
+configure your applications. Symfony lets you choose between YAML, XML and PHP
+and throughout the Symfony documentation, all configuration examples will be
+shown in these three formats.
 
-.. sidebar:: Configuration Formats
+There isn't any practical difference between formats. In fact, Symfony
+transforms all of them into PHP before running the application, so there's not
```
Comment (wouterj) [inline configuration.rst]: "transforms and caches" maybe, as transforming sounds like a performance impact?

File: configuration.rst [hunk 2]
```diff
@@ -156,26 +181,30 @@ package, a ``locale`` parameter is added  to ``config/services.yaml``:
...
+        // the parameter name is an arbitrary string (the 'app.' prefix is recommended
+        // to better differentiate your parameters from Symfony parameters).
+        // the parameter value can be any valid scalar (numbers, strings, booleans, arrays)
+        $container->setParameter('app.admin_email', 'something@example.com');
         // ...
 
-This parameter is then referenced in the framework config in
-``config/packages/translation.yaml``:
+Once defined, you can reference this parameter value from any other
+configuration file using a special syntax: wrap the parameter name with two
```
Comment (wouterj) [inline configuration.rst]:
```suggestion
configuration file using a special syntax: wrap the parameter name in two
```

File: configuration.rst [hunk 3]
```diff
@@ -186,111 +215,353 @@ This parameter is then referenced in the framework config in
                 https://symfony.com/schema/dic/symfony/symfony-1.0.xsd">
 
             <!-- any string surrounded by two % is replaced by that parameter value -->
-            <framework:config default-locale="%locale%">
+            <some-package email-address="%app.admin_email%">
```
Comment (wouterj) [inline configuration.rst]: This is never valid XML, let's do something like this:

```xml
<some-package:config email-address="%app.admin_email%">
    <!-- ... -->
</some-package:config>
```

File: configuration.rst [hunk 4]
```diff
@@ -186,111 +215,353 @@ This parameter is then referenced in the framework config in
...
+
+Symfony applications come with a file called ``.env`` located at the project
+root directory. This file is used to define the value of environment variables
+and it's explained in detail :ref:`later in this article <config-dot-env>`.
+
+Open the ``.env`` file and edit the value of the ``APP_ENV`` variable to change
+the environment in which the application runs. For example, to run the
+application in production:
+
+```bash
```
Comment (wouterj) [inline configuration.rst]: whoa, markdown syntax 😉

File: configuration.rst [hunk 5]
```diff
@@ -186,111 +215,353 @@ This parameter is then referenced in the framework config in
...
+    learn about how to use these configuration parameters in services and controllers.
+
+.. index::
+   single: Environments; Introduction
+
+.. _page-creation-environments:
+.. _page-creation-prod-cache-clear:
+.. _configuration-environments:
+
+Configuration Environments
```
Comment (wouterj) [inline configuration.rst]: I think somewhere in this article, we'll have to explain that all configuration does is actually configuring the services in Service Container. And by configuring the services, all framework behaviour can be changed. This, imo, makes the user understand configuration and environments better (also in combination with service parameters explained above).

File: configuration.rst [hunk 6]
```diff
@@ -186,111 +215,353 @@ This parameter is then referenced in the framework config in
...
+before going to production:
 
-For more information about parameters - including how to reference them from inside
-a controller - see :ref:`service-container-parameters`.
+#. Create a configuration directory with the same name as the environment (in
+   this case, ``config/packages/staging/``);
+#. Add the needed configuration files in ``config/packages/staging/`` to
+   define the behavior of the new environment. Symfony loads first the files in
+   ``config/packages/*.yaml``, so you must only configure the differences with
+   those files;
```
Comment (wouterj) [inline configuration.rst]: Maybe add a tip that, if the configuration is equal to some other env, using symlinks is a great idea? I.e. in our staging environment, almost everything is production settings, except from `framework.debug` and monolog. So all other `config/packages/prod/*.yaml` are symlinked to `config/packages/staging/*.yaml`

File: configuration.rst [hunk 7]
```diff
@@ -186,111 +215,353 @@ This parameter is then referenced in the framework config in
...
+.. _config-env-vars:
+
+Configuration Based on Environment Variables
+--------------------------------------------
+
+Using `environment variables`_ (or "env vars" for short) is a common practice to
+configure sensitive options such as credentials and passwords. Instead of
+defining the values of those options in the configuration files, you add a
+reference to an environment variable which will be resolved at runtime (only
+once per request, to not impact performance).
```
Comment (wouterj) [inline configuration.rst]: I'm against putting this recommendation (or even hinting at doing stuff like this) in the documentation. Using env vars to store sensitive options is a *bad* idea (see i.e. https://www.webfactory.de/blog/storing-secrets-for-symfony-applications ).

Instead: Env vars shine for settings that change per environment, e.g. a DATABASE_URL (which isn't sensitive).

We can then add a section explaining some of the env var processors (i.e. `file:` or `json:`) to show a better practice to store sensitive data.

File: configuration/front_controllers_and_kernel.rst [hunk 8]
```diff
@@ -120,6 +120,82 @@ new kernel.
...
+        # config/packages/twig.yaml
+        twig:
+            debug: '%kernel.debug%'
+
+    .. code-block:: xml
+
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xmlns:doctrine="http://symfony.com/schema/dic/doctrine"
```
Comment (wouterj) [inline configuration/front_controllers_and_kernel.rst]:
```suggestion
            xmlns:twig="http://symfony.com/schema/dic/twig"
```

**Review by wouterj — CHANGES_REQUESTED** [review]
Great work! Did a quick review, it's to big to review very thoroughly :)

**Comment by javiereguiluz** [comment]
Wouter, thanks for your review. I think you were right at everything, so I made some fixes and improvements.

**Review by wouterj — APPROVED** [review]
Looks great!

**Pull request merged**